### PR TITLE
index street without admin

### PR DESCRIPTION
**Pull request by xlqian**
http://jira.canaltp.fr/browse/NAVP-557

### CODE REVIEW

File: src/bin/osm2mimir.rs [hunk 1]
```diff
@@ -40,6 +40,8 @@ use std::collections::HashSet;
 use mimirsbrunn::rubber::Rubber;
 
 pub type AdminsVec = Vec<mimirsbrunn::Admin>;
+pub type StreetsVec = Vec< mimirsbrunn::Street>;
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: useless space

File: src/bin/osm2mimir.rs [hunk 2]
```diff
@@ -160,15 +168,59 @@ fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec
...
+                way.tags.get("highway").map_or(false, |x: &String| !x.is_empty()) &&
+                way.tags.get("name").map_or(false, |x: &String| !x.is_empty())
+            }
+            _ => false,
+        }
+    };
+
+    let mut streets = StreetsVec::new();
+
+    if let Result::Ok(ref objects) = osmpbfreader::get_objs_and_deps(pbf, is_valid_street) {
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: I'm far from being an expert, but I think you can write this in a bit more rusty way with `map` `collect` and `and_then`

I can help you with this if you want

File: src/bin/osm2mimir.rs [hunk 3]
```diff
@@ -179,10 +231,8 @@ fn main() {
...
-    let res = administrative_regions(&args.flag_input, levels);
-    match index_osm(&args.flag_connection_string, &res) {
-        Err(e) => panic!("failed to index osm because: {}", e),
-        Ok(nb) => info!("Adminstrative regions: {}", nb),
-    }
-
+    let mut parsed_pbf = parse_osm_pbf(&args.flag_input);
+    let admins = administrative_regions(&mut parsed_pbf, levels);
+    let streets = streets(&mut parsed_pbf);
+    index_osm(&args.flag_connection_string, &admins, &streets);
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: are you forcer to take `streets` and `admins` by ref ? can't you move them ?

File: src/bin/osm2mimir.rs [hunk 4]
```diff
@@ -83,13 +85,19 @@ impl AdminMatcher {
     }
 }
 
-fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec {
+
+fn parse_osm_pbf(path: &String) -> ParsedPbf {
+    let path = std::path::Path::new(&path);
+    match std::fs::File::open(&path) {
+        Err(e) => panic!("failed to parse file {:?} because: {}", &path, e),
```
Comment (kinnou02) [inline src/bin/osm2mimir.rs]: Not sure it's better than an ```unwrap```

File: src/bin/osm2mimir.rs (hunk 3 above)
Comment (kinnou02) [inline src/bin/osm2mimir.rs]: we should index the admins, then index the street, it will help with the memory

File: src/bin/osm2mimir.rs [hunk 5]
```diff
@@ -160,15 +165,86 @@ fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec
...
 }
 
-fn index_osm(es_cnx_string: &str, admins: &AdminsVec) -> Result<u32, rs_es::error::EsError> {
+fn streets(pbf: &mut ParsedPbf) -> StreetsVec {
+
+    let is_valid_street = |obj: &osmpbfreader::OsmObj| -> bool {
+        match *obj {
+            osmpbfreader::OsmObj::Way(ref way) => {
+                way.tags.get("highway").map_or(false, |x: &String| !x.is_empty()) &&
+                way.tags.get("name").map_or(false, |x: &String| !x.is_empty())
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: nice :+1:

File: src/bin/osm2mimir.rs [hunk 6]
```diff
@@ -160,15 +165,86 @@ fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec
...
+                rel.tags.get("type").map_or(false, |v| v == "associatedStreet")
+            }
+            _ => false,
+        }
+    };
+
+    let mut street_map = osmpbfreader::get_objs_and_deps(pbf, is_valid_street).unwrap();
+    let relation_map = osmpbfreader::get_objs_and_deps(pbf, is_associated_street_relation).unwrap();
+
+    for (_, rel_obj) in &relation_map {
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: can you add a comment about what you want to do ?

Comment (xlqian) [inline src/bin/osm2mimir.rs]: yep

File: src/bin/osm2mimir.rs (hunk 4 above)
Comment (xlqian) [inline src/bin/osm2mimir.rs]: yep

**Comment by antoine-de** [comment]
Ok for me with the deactivation

File: src/bin/osm2mimir.rs [hunk 7]
```diff
@@ -83,13 +85,16 @@ impl AdminMatcher {
     }
 }
 
-fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec {
+
+fn parse_osm_pbf(path: &String) -> ParsedPbf {
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: &str

File: src/bin/osm2mimir.rs (hunk 5 above)
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Useless type info.

File: src/bin/osm2mimir.rs [hunk 8]
```diff
@@ -160,15 +165,94 @@ fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec
...
+        match *obj {
+            osmpbfreader::OsmObj::Relation(ref rel) => {
+                rel.tags.get("type").map_or(false, |v| v == "associatedStreet")
+            }
+            _ => false,
+        }
+    };
+
+    let mut street_map = osmpbfreader::get_objs_and_deps(pbf, is_valid_street).unwrap();
+    let relation_map = osmpbfreader::get_objs_and_deps(pbf, is_associated_street_relation).unwrap();
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: You'll read a lot of time the pbf.

Comment (xlqian) [inline src/bin/osm2mimir.rs]: I have to get these two maps separately... i agree with you but i don't know how to get them in one shot

File: src/bin/osm2mimir.rs [hunk 9]
```diff
@@ -160,15 +165,94 @@ fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec
...
+
+    // Sometimes, streets can be devided into several "way"s that still have the same street name.
+    // The reason why a street is devided may be that a part of the street become a bridge/tunne/etc.
+    // In this case, a "relation" tagged with (type = associatedStreet) is used to group all these "way"s.
+    // In order not to have doublons in autocompleion, we should keep only one "way" in the relation
+    // and remove all the rest way whose street name is the same.
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: If we want to compute the centroid, we'll need every ways, not only one.

File: src/bin/osm2mimir.rs [hunk 10]
```diff
@@ -282,15 +288,93 @@ fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec
...
+    for (_, rel_obj) in &relation_map {
+        if let &osmpbfreader::OsmObj::Relation(ref rel) = rel_obj {
+            let mut found = false;
+            for ref_obj in &rel.refs {
+                if let &osmpbfreader::OsmId::Way(_) = &ref_obj.member {
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Maybe we can add a is_way et al to OsmId and OsmObj.

File: src/bin/osm2mimir.rs [hunk 11]
```diff
@@ -282,15 +288,93 @@ fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec
...
+                    if !found {
+                        found = true;
+                        continue;
+                    };
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Useless ;

File: src/bin/osm2mimir.rs [hunk 12]
```diff
@@ -282,15 +288,93 @@ fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec
...
+                    street_map.remove(&ref_obj.member);
+                };
+            }
+        };
+    }
+
+    street_map.iter()
+              .filter_map(|(osm_id, obj)| {
+                  if let &osmpbfreader::OsmObj::Way(ref way) = obj {
+                      if let &osmpbfreader::OsmId::Way(ref way_id) = osm_id {
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]:
```rust
use osmpbfreader::{OsmId, OsmObj};
if let (&OsmId::Way(ref way_id), &OsmObj(ref way)) = (osm_id, obj) {
```

File: src/bin/osm2mimir.rs [hunk 13]
```diff
@@ -282,15 +288,93 @@ fn administrative_regions(filename: &String, levels: HashSet<u32>) -> AdminsVec
...
+                          way.tags.get("name").and_then(|way_name| {
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: map is better than and_then in this case.

File: src/bin/osm2mimir.rs (hunk 12 above)
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: ah yes, great!

File: src/bin/osm2mimir.rs [hunk 14]
```diff
@@ -301,10 +385,19 @@ fn main() {
...
+    index_settings(&args.flag_connection_string);
+    index_admin(&args.flag_connection_string, admins); 	
+    
+    let mut optinoal_streets: Option<StreetsVec> = None;
+    if args.flag_import_way  {
+        optinoal_streets = Some(streets(&mut parsed_pbf));    
+    }
+    
+    if let Some(streets) = optinoal_streets {
+	index_ways(&args.flag_connection_string, streets);
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Why not just do the job on one if ? This option dance is difficult to read.

File: src/objects.rs.in [hunk 15]
```diff
@@ -31,6 +31,7 @@
 
 // Note: this file is needed to use serde in rust stable
 // cf mimirsbrunn/build.rs for explanations
+use std::option::Option;
```
Comment (TeXitoi) [inline src/objects.rs.in]: You need that ? Strange it should be imported by the prelude.

File: src/bin/osm2mimir.rs [hunk 16]
```diff
@@ -42,23 +42,27 @@ use mimirsbrunn::rubber::Rubber;
 use mimirsbrunn::objects::{Polygon, MultiPolygon};
 
 pub type AdminsVec = Vec<mimirsbrunn::Admin>;
+pub type StreetsVec = Vec<mimirsbrunn::Street>;
+pub type ParsedPbf = osmpbfreader::OsmPbfReader<std::fs::File>;
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: That's not parsed, that's more a parser.

File: src/bin/osm2mimir.rs [hunk 17]
```diff
@@ -42,23 +42,27 @@ use mimirsbrunn::rubber::Rubber;
...
     flag_level: Vec<u32>,
     flag_connection_string: String,
+    flag_import_way: bool
 }
 
 static USAGE: &'static str = "
 Usage:
     osm2mimir --help
-    osm2mimir --input=<file> [--connection-string=<connection-string>] --level=<level>...
+    osm2mimir --input=<file> [--connection-string=<connection-string>] --level=<level> [--import-way]...
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: The ... Are badly placed.

File: src/bin/osm2mimir.rs (hunk 12 above)
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: And there is the id in the way, you don't need to use the OsmId.

File: src/bin/osm2mimir.rs (hunk 9 above)
Comment (xlqian) [inline src/bin/osm2mimir.rs]: can we calculate the centroid in the loop of https://github.com/CanalTP/mimirsbrunn/pull/22/files#diff-8be6b8fd629e2b03d7a46cd92d72be2bR323?

File: src/bin/osm2mimir.rs [hunk 18]
```diff
@@ -42,23 +42,27 @@ use mimir::rubber::Rubber;
...
 Usage:
     osm2mimir --help
-    osm2mimir --input=<file> [--connection-string=<connection-string>] --level=<level>...
+    osm2mimir --input=<file> [--connection-string=<connection-string>] --level=<level> ... [--import-way]
 
 Options:
     -h, --help            Show this message.
     -i, --input=<file>    OSM PBF file.
     -l, --level=<level>   Admin levels to keep.
+    -w, --import-way      By default, streets are not imported
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: the comment does not seems synchronized

File: src/bin/osm2mimir.rs [hunk 19]
```diff
@@ -301,10 +387,19 @@ fn main() {
...
+    let mut parsed_pbf = parse_osm_pbf(&args.flag_input);
+
+    let admins = administrative_regions(&mut parsed_pbf, levels);
+
+    index_settings(&args.flag_connection_string);
+    index_admin(&args.flag_connection_string, admins);
+
+    let mut optinoal_streets: Option<StreetsVec> = None;
+    if args.flag_import_way {
+        optinoal_streets = Some(streets(&mut parsed_pbf));
```
Comment (antoine-de) [inline src/bin/osm2mimir.rs]: why don't you put all in the if ?

you can rewrite it: 

```
if args.flag_import_way {
        let streets = streets(&mut parsed_pbf);
        index_ways(&args.flag_connection_string, streets);
}
```

**Pull request merged**